### ISSUE TITLE
internal/shader: possible breaking changes in `position` in Kage between v2.4 and v2.5

### ISSUE THREAD

**Issue by hajimehoshi** [hajimehoshi/ebiten#2639]
### Ebitengine Version

v2.5.2

### Operating System

- [ ] Windows
- [ ] macOS
- [ ] Linux
- [ ] FreeBSD
- [ ] OpenBSD
- [ ] Android
- [ ] iOS
- [ ] Nintendo Switch
- [ ] Xbox
- [ ] Web Browsers

### Go Version (`go version`)

n/a

### What steps will reproduce the problem?

As Ebitengine v2.5 uses texture atlases not only for rendering sources but also for rendering destinations. This means that the argument `position` in `Fragment` can be different from v2.4.

In order to keep compatibility, Ebitengine might have to stop texture-atlasing for custom shaders.

### What is the expected result?

The same behavior as v2.4.

### What happens instead?

A possibly different behavior from v2.4.

### Anything else you feel useful to add?

/CC @Zyko0 @SolarLune @tinne26 as heavy-Kage users

**Comment by hajimehoshi** [comment]
Also, with anti-aliasing, the destination region is doubled. This is still surprising, but this is NOT a breaking change as anti-aliasing doesn't exist in v2.4.

**Comment by hajimehoshi** [comment]
Let's think more whether this should be fixed and backported to v2.5, or this doesn't have to be fixed.

**Comment by SolarLune** [comment]
We discussed this on the Discord, but yes, I believe it probably would be best to simply turn off texture atlasing when using shaders. That way `position` and `texCoord` do what they're supposed to do.

The only real alternative would be to fix it so that `position` and `texCoord` always work properly; if that's not realistic, then just turning off atlasing seems to be fine.

**Comment by hajimehoshi** [comment]
We don't have to fix texCoord, right?

> so that position and texCoord always work properly;

What does working properly mean here? At least, I think they are working 'properly' in v2.5, but in a little different way of v2.4.

> EDIT: To be clear, I mean that Ebitengine should automatically turn off atlasing for textures when using them as source or destination images with shaders.

Now every rendering uses shaders, so this is not feasible to turn antialiasing off for every shader. My suggestion is to turn off 'atlasing' when a shader uses `position` parameter.

**Comment by hajimehoshi** [comment]
Also, whether turning off anti-aliasing is not the topic of this issue. As I said, the behavior with anti-aliasing is NOT a breaking change.

**Comment by SolarLune** [comment]
> What does working properly mean here? At least, I think they are working 'properly' in v2.5, but in a little different way of v2.4.

"Working properly" depends on how it's defined, but I would call something "working properly" when it is working as intended and is consistent.

In other words, texCoord should always point to the interpolated UV values of the fragment. If you're rendering a texture and the vertices' UV values range from 0 to 1, then the texCoord values for the rendered fragments should range from 0 to 1. Sampling textures using texture coordinate values should work directly, as well.

`position` should, similarly, be consistent and predictable. It should either be the position of the pixel on the screen, or the position of the fragment onscreen as a percentage of the screen size. It should never be unknown (which is what it is, currently).

> Also, whether turning off anti-aliasing is not the topic of this issue. As I said, the behavior with anti-aliasing is NOT a breaking change.

Understood, I don't have any complaints about the anti-aliasing system and wasn't referring to it.

> Now every rendering uses shaders, so this is not feasible to turn antialiasing atlasing off for every shader. My suggestion is to turn off 'atlasing' when a shader uses position parameter.

It's up to you, but I meant just turning off atlasing when a custom shader is used at all. The default / built-in shaders aren't generally a problem.

**Comment by hajimehoshi** [comment]
> In other words, texCoord should always point to the interpolated UV values of the fragment. If you're rendering a texture and the vertices' UV values range from 0 to 1, then the texCoord values for the rendered fragments should range from 0 to 1. Sampling textures using texture coordinate values should work directly, as well.

I'm afraid I'm against it.

1. This might cause a little overhead.
2. texCoord 1 doesn't make sense. This doesn't always match with lower-right point.
3. 0 to 1 is not intuitive when the image is sub-image.
4. Adopting pixels (#1431) for texCoord sounds better idea to me.

> position should, similarly, be consistent and predictable. It should either be the position of the pixel on the screen, or the position of the fragment onscreen as a percentage of the screen size (which is probably the better option of the two, as this makes it match texCoord / UV values more closely, which I believe is both correct and is also how GLSL does it). It should never be unknown (which is what it is, currently).

So you mean the upper-left point should always be (0, 0)?

**Comment by SolarLune** [comment]
> I'm afraid I'm against it.

1. A little overhead is fine when we're talking about custom shaders, I think - it's not a feature everyone would use, and it already should be offloading work from the CPU to the GPU, or only doing what the GPU can do. If it was a bit slower to send data to the GPU because of conversions, I think it would be worth the tradeoff (though doing some testing and seeing how much of a tradeoff would be nice).
2. It should always match with the lower-right point. `imageSrcNAt` should work with a normalized texture coordinate (0 to 1) range.
3. The range being 0 to 1 is more intuitive than the current alternative, which is that the range is unknown, essentially, and that you have to use functions to convert it to a usable, normalized range.e
4. It's up to you. I would probably rather it worked like other shading languages, but if it's easier to you to do, then that's fine.

>  So you mean the upper-left point should always be (0, 0)?

Yes.

___

I don't mind too much what you end up choosing for this, but I just want it to be simple, and for `position` and `texCoord` to, ideally, be usable directly without having to convert or change them. Oh, and if resolving this also allows for using images of varying sizes in shaders, that would also be useful.

**Comment by hajimehoshi** [comment]
> That's how texture sampling works already for GLSL (and I'd imagine other shader languages), though, if I'm correct...? 

Ebitengine treats texture atlases for both source and destination images. So imageSrcNAt would requrie some work if we want to use a normalized range. I don't want to introduce this overhead.

**Comment by hajimehoshi** [comment]
> I might have misspoke when I said "normalized" - I meant that it ranges from 0 at the top / left to 1 at the bottom / right.

Even without atlases, the bottom-right texel is not (1, 1) in general since a texture size is power of 2 while an image size might not be. And even without atlases, (though this is an implementation detail,) the left-upper texel might not be (0, 0) to be exact as there is a thin border. 

Of course, with atlases, the range is no longer (0, 0)-(1, 1).

**Comment by tinne26** [comment]
I agree with SolarLune that `position` and `texCoord` should be as predictable and have as few special cases as possible. That being said, I also understand why this may be hard or even impossible.

For me, the ideal situation would be for `position` to keep working in pixels but Ebitengine handles atlas offsets internally, preventing that from affecting the user. I don't know whether this is possible, but my main point here is that allowing `position` to work with pixels and predictably means that `position` is a very practical input parameter. In fact, the whole intro tutorial of kage-desk revolves around this. Working with pixels is very valuable even on hiDPI-aware programs, as many sprites and effects can still be done at a specific size and only upscaled later as required.

So, this is my first main point: `position` can be useful, and it should be. The current v2.5 underspecified behavior goes pretty heavily against this. And I would definitely not want `position` to disappear.

Of course, having to do shaders outside atlases may become a performance problem, but I want to highlight that the fact that SolarLune used a separate unmanaged offscreen to get around the quirks affecting `position` should not be treated as a story of success (e.g. "ok something broke but it was quickly fixed"). The fix may not take a lot of code, but requires knowledge about Ebitengine internals and requires complicated explanations. I don't think this is ok for the average user.

I think a good next step is writing down all the tricky cases, because they are too many and they are too tricky. It's not just that we don't know how to make `position` simpler, it's that no one has a clear view of what happens with the shader parameters on all the permutations of API calls and configurations. E.g.: subimages can affect shader inputs, source image atlases can affect shader inputs, target atlases can affect shader inputs, `DrawRectShader` includes `GeoM` which can affect shader inputs, antialias can affect shader inputs, etc. This also kinda applies to `texCoord` to some degree. I think having such a list of "special cases" with the behavior detailed is a requirement before we can continue discussing this. Then we can go point by point saying which special cases seem acceptable and which ones don't.

For example, if I try to think about the case where we pass multiple images and they are all the same size but are on different atlases and are subimages with different origins, my brain simply short-circuits. Like, yeah, maybe nothing weird will happen, but in v2.5, I wouldn't dare say what will happen without actually trying it. Like, I see what you guys are discussing and I'm like "ok, I can't even say anything yet because I don't even know where we are right now".

**Comment by Zyko0** [comment]
Agree with both @tinne26 and @SolarLune on the user experience side of things
And I think @hajimehoshi also agrees, but is replying in terms of implementation

1. We need a way to measure the "overhead" we would get from `imageSrcNAt` being magic (able to treat 0-1 range coordinates). Because as mentioned earlier, whenever we start working with a shader, it is in most cases to be able to play with coordinates and manipulate them. So we're almost everytime having this overhead ourselves by transforming coords back, we could also just transform them once when about to fetch texture (getting it as 0-1, and doing *size+origin after)
2. We should check if there's a workaround to the sub-image issue (https://github.com/hajimehoshi/ebiten/issues/1431#issuecomment-1507034265), because atm it doesn't seem feasible at all because of it.. :/

For 2, if the size needs to be changed at uniform level, therefore involves breaking batching, I don't think it can't be done (or at least not this way)

**Comment by SolarLune** [comment]
> So, this is my first main point: position can be useful, and it should be. The current v2.5 underspecified behavior goes pretty heavily against this. And I would definitely not want position to disappear.

Definitely, `position` is highly useful in shaders.

> I want to highlight that the fact that SolarLune used a separate unmanaged offscreen to get around the quirks affecting position should not be treated as a story of success (e.g. "ok something broke but it was quickly fixed"). The fix may not take a lot of code, but requires knowledge about Ebitengine internals and requires complicated explanations (which are basically: let's create a situation where we have the same context as we used to in v2.4). I don't think this is ok for the average user.

Agreed, yeah - creating an unmanaged image isn't hard, of course, but it shouldn't be "the fix" to make `position` work consistently or "knowably". Ideally, the user shouldn't have to think about it.

Then again, if making a texture unmanaged is the only way to have `position` work reliably, then that would be fine, I think, if it were properly documented (or, better yet, if it was done automatically).

> I think a good next step is writing down all the tricky cases, because they are too many and they are too tricky.

This probably is a good way forward - defining what the issue is, clearly, should help make things simpler to approach. The alternative would be to start from the end and work backwards - what do we want `position` and `texCoords` to represent, and how do we get that?

>  And I think @hajimehoshi also agrees, but is replying in terms of implementation

Yes, you're right - thanks for working with us on this, @hajimehoshi - I know it's not easy, especially when we aren't (or at least I am not) really familiar with the internals of Ebiten's workings that would make this or other issues difficult or even impossible to resolve.

**Comment by hajimehoshi** [comment]
Just a quick reply: what do you think about using pixels instead of texels (#1431)? If this is the way to go, we don't have to think about 'normalizing' to 0-1.

**Comment by hajimehoshi** [comment]
> I think a good next step is writing down all the tricky cases, because they are too many and they are too tricky. It's not just that we don't know how to make position simpler, it's that no one has a clear view of what happens with the shader parameters on all the permutations of API calls and configurations. E.g.: subimages can affect shader inputs, source image atlases can affect shader inputs, target atlases can affect shader inputs, DrawRectShader includes GeoM which can affect shader inputs, antialias can affect shader inputs, etc. This also kinda applies to texCoord to some degree. I think having such a list of "special cases" with the behavior detailed is a requirement before we can continue discussing this. Then we can go point by point saying which special cases seem acceptable and which ones don't.

I'll try to write a Google doc to summrize the current situation and my opinion. Stay tuned.

**Comment by hajimehoshi** [comment]
@tinne26 I'm still writing the document, but I hope the current summery of Ebitengine's implementation would help you to understand:

https://docs.google.com/document/d/1lJTKxMqyyO-LxNS4KmuNakQwlhdBgzkmP2LNwpJk5bg/edit#bookmark=id.eglkzn1frx1r

**Comment by hajimehoshi** [comment]
So, do you all agree that converting`position` and `texCoord` to [0, 1] is not feasible? One of the reasons is that this forces to use uniform variables in a shader program and then affects batching. As I suggested in the document, utility functions to normalize values should be fine.

**Comment by tinne26** [comment]
The following is my "ideal model".

We already discussed this somewhere else, but I'm repeating it here for anyone else interested in the issue.

There are some implementation and compatibility hurdles that make it not viable for v2 and probably also impractical or undesirable even for v3, but I think it's useful as a "what would be nice to have" reference.

My model considers a `func Fragment() vec4` with no inputs. The main goal of the model is to offer two sets of functions: one predictable and easy to use that doesn't care much about performance and batching breaks, and another more "low level". The syntax is given as pseudo-code, not final.

The "easy to use" set of functions:
```Golang
TargetCoords() vec4 // in pixels, =~ `position` - atlas_offsets - border_offsets - subimage_offsets
SubTargetCoords() vec4 // in pixels, =~ `position` - atlas_offsets - border_offsets
TargetSize() vec2 // in pixels
Textures[N].Mapping() vec2 // src image coords mapped to the current fragment, in pixels
Textures[N].At(vec2) vec4 // in pixels. refers to source images.
Textures[N].UnsafeAt(vec2) vec4
Textures[N].Size() vec2 // all would return the same
Attributes[N] float // up to 16 values per vertex
```
The "low-level" set of functions:
```Golang
RawTargetCoords() vec4 // in pixels. =~ current `position`
Textures[N].RawTexelMapping() vec2 // removes some magic
Textures[N].AtRawTexels(vec2) vec4 // or pixels if that's equally or more efficient
Textures[N].RawSizeInTexels() vec2 // or pixels if that's equally or more efficient
Textures[N].UnsafeAtRawTexels(vec2) vec4 // or pixels if that's equally or more efficient
```

Notice that actual functionality and calls are almost the same that we already have or that Hajime's proposal adds (most parts of this proposal could be implemented through a pre-processor, e.g., on `kage-desk/display`). The only notable points are:
- Clear separation of easy to use and low-level APIs.
- Removing input arguments from `Fragment()` is not a direct goal of my model, but it spares the average user from having to deal with input arguments that require understanding internal intricacies of Ebitengine (e.g. atlases and borders) to be used properly.
- Currently in kage there's only one `imageSrcRegionOnTexture()` function and return value, which is a lie / hack (as different source images can have different origins) that's compensated later on `imageSrcNAt()`. I dislike this and prefer the more explicit and magic-free `Textures[N].RawTexelMapping()`. For those wondering, `Textures[N].Mapping()` would return the same value for all source textures instead (could also be written as `TextureMapping()` or something).
- Pixels are preferred and prioritized over texels. This matches Hajime's intent on [#1431](https://github.com/hajimehoshi/ebiten/issues/1431) and I think we are all pretty much on the same page on this.
- Generalization of `Attributes` is also suggested (see [#2640](https://github.com/hajimehoshi/ebiten/issues/2640)). This is not a core part of this model, but it's motivated by the fact that many shaders operate with simple quads, and when you only have four vertices, being able to pass some values that would typically be hidden uniforms as attributes can actually make sense (not too much overhead, no batching breaking). We would still need to expose something like `Image.AtlasOrigin()` on the CPU side to make this really useful, but that's a story for another day.

**Comment by hajimehoshi** [comment]
@tinne26 thanks. Could you file a new issue or a discussion and move your comment there? Now there are a lot of topics here and I'd like to focus on solving the issue in v2 without breaking changes here.

**Comment by hajimehoshi** [comment]
For normalizing, I created another issue #2644

**Comment by SolarLune** [comment]
> So, do you all agree that automatic convertingposition and texCoord to [0, 1] is not feasible? One of the reasons is that this forces to use uniform variables in a shader program and then affects batching. As I suggested in the document, utility functions to normalize values should be fine.

If it's not feasible or would hurt performance, then it isn't the ideal solution. You know better than us if it's possible, of course, but I'll follow you lead on that one.

___

I think switching to pixels for `position` sounds like a reasonable compromise that wouldn't be hard to deal with for the user, while producing reliable results and being feasible to implement internally (at least, from my point of view).

I do, though, feel like pixel mode should be the default without a directive; the texel mode can be the directive. If shaders break between the previous release of ebiten and the current one, and users have to update anyway, it might as well be updated in the most future-proof way possible.

**Comment by hajimehoshi** [comment]
> I think switching to pixels for position sounds like a reasonable compromise that wouldn't be hard to deal with for the user, while producing reliable results and being feasible to implement internally (at least, from my point of view).

`position` is already in pixels. The thing that will be changed is `texCoord` from texel to pixel.

**Comment by hajimehoshi** [comment]
> I do, though, feel like pixel mode should be the default without a directive; the texel mode can be the directive. If shaders break between the previous release of ebiten and the current one, and users have to update anyway, they might as well update in the most future-proof way possible, and not to have the default be broken.

I'm afraid I disagree with this. For example, such a breaking change would make it extremely difficult for users to experiment multiple versions of Ebitengine to find some issues. Users are sometimes forced to use old versions of Ebitengine when a new Ebitengine has some issues.

**Comment by Zyko0** [comment]
> I'm afraid I disagree with this. For example, such a breaking change would make it extremely difficult for users to experiment multiple versions of Ebitengine to find some issues. Users are sometimes forced to use old versions of Ebitengine when a new Ebitengine has some issues.

The fact that "Kage" is a separate language, and quite a distinct fonctionnality, that shader files are "assets" makes this a fair point to me.
Otherwise, I would favor breaking change too.

Thinking about a "mode" and a "directive" I'd still prefer going with a "version" though 👀 (where default/legacy are v1 and v2 only supports pixels).
Although it sounds like semantics, it allows to slowly deprecate the old version with "texels".
While if you have a "mode", you're officially supporting two Kage shader modes for the current set of features, and the potential new features to come.

Also, whether you go with a mode or a version, I think we still need a way to encourage users to write their new shaders in "pixel units".
Which again, I think is more encouraging by specifying that the shader is "v2" as opposed to "//kage-mode:pixels".
Users will probably be more willing to set the latest Kage version in their file, rather than pick a new "mode".

If the fact that users might be confused about what Kage shader is in pixels or in texels (just from a "version"), I do not think that's an issue, the fact that it breaks with the wrong directive does, and the reason being that units are in texels or pixels could be an internal detail.

**Comment by hajimehoshi** [comment]
> Thinking about a "mode" and a "directive" I'd still prefer going with a "version" though 👀 (where default/legacy are v1 and v2 only supports pixels).

Users would be confused with versions as it would be hard to know which feature works and which is not. So I prefer having a specific mode with an explicit name.

**Comment by SolarLune** [comment]
> position is already in pixels. The thing that will be changed is texCoord from texel to pixel.

Ah, my mistake; thanks for the correction. Looking at the proposal, it looks like this would also affect the functions that currently take texels, which is good.

> > I do, though, feel like pixel mode should be the default without a directive; the texel mode can be the directive. If shaders break between the previous release of ebiten and the current one, and users have to update anyway, they might as well update in the most future-proof way possible, and not to have the default be broken.
> 
> I'm afraid I disagree with this. For example, such a breaking change would make it extremely difficult for users to experiment multiple versions of Ebitengine to find some issues. Users are sometimes forced to use old versions of Ebitengine when a new Ebitengine has some issues.

While a compiling directive would allow people to easily swap between different versions of Ebitengine more easily, non-trivial shaders that work in 2.4 would still be broken by default in the latest version, so I don't think you get very much for your trouble there, unless you were also going to disable atlasing when shaders are used.

Most Ebitengine users probably either don't use shaders or are aware of the problem (like us), so I don't think there's very many people who would be negatively impacted, but it's up to you.

**Comment by hajimehoshi** [comment]
One idea is stopping atlasing for destination textures when the mode is the texel mode (including v2.5), so that we can fix the slightly broken compability issue (this issue). On the contrary, when the mode is the pixel mode, we enable the atlasing for destination textures.

**Comment by tinne26** [comment]
I agree that reverting v2.5 to behave like v2.4 (no atlasing) is the most immediate patch for this issue.

This doesn't change the big problem, which is that input parameters are too hard to use, basically in any mode, and any tutorial, even with the new proposed functions to make life easier, kinda requires an initial setup and explanation and a "basically, you need to use these functions anyway because otherwise this is too difficult to use yet". To me, this means we need to discuss more in depth where we want to go and figure out how to make it nicer for the average user, even if it requires a `//v2`, or a later breaking change or some other directive. But before adding any directive like that, I think we need to discuss all this in a separate issue to make the changes really future proof. I think your proposal with the pixel mode directive and the new functions "works", but it's not as easy to use or pleasant as it should be in the long run (imo). So I'm skeptical of starting to add compiler directives to "quick-fix" multiple of the currently open issues in a way that feels somewhat hacky.

I vote for reverting to v2.4 behavior and opening a new issue to discuss the model we want to go towards. Unless you really feel that pixel mode and the new proposed functions are actually satisfactory in the long run, of course.

**Comment by hajimehoshi** [comment]
I fixed the example in v2.5:

* 96f9ea1b2e8ade7b8eaff0341c97f114a665985c
* 038cacf97f5adac8589be25f8063e0596726c97f

As long as no more people (other than the people here) complain about not reverting the behavior, I'll keep the current behavior as it is...

**Comment by hajimehoshi** [comment]
Another idea to determine the behavior is to try to use Ebitengine v2.5 for the games in the game jam 2022 and see what's going on.

**Comment by hajimehoshi** [comment]
> As long as no more people (other than the people here) complain about not reverting the behavior, I'll keep the current behavior as it is... I feel like the number of people using Kage actively is not so big.

I'll close this issue.
